plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Escrow.hs

```diff
@@ -273,11 +275,11 @@ refund
     :: ( HasUtxoAt s
        , HasWriteTx s )
     => EscrowParams 
-    -> TxId
+    -> PubKey
     -> Contract s RefundResult
-refund escrow txId = do
+refund escrow pk = do
     unspentOutputs <- utxoAt (escrowAddress escrow)
-    let flt TxOutRefOf{txOutRefId} _ = txId Haskell.== txOutRefId
+    let flt _ txOut = Ledger.txOutData txOut == Just (DataScript (PlutusTx.toData pk))
```

Comment:
Interesting thing here: we probably want to *decode* the data and do equality on the decoded values (under the `Maybe` is fine!), since the decoded type might have a different notion of equality!

Reply:
Ahhhh. That's a good point.  We're still claiming that on-chain scripts could potentially be written in other languages (maybe with different notions of equality), and when we create an output we generally won't know who's going to be consuming it and its `Data`.  Even different validators written in Plutus might decode `Data` objects in different ways, or treat them differently once they're decoded (there's no reason why something encoded with the `List` constructor can't be decoded into a random-aceesss array/vector, for example).  Are we implicitly assuming that a Data "script" has a particular semantics which must be respected during validation?  I suppose an output's owner still gets to inspect the validator in advance, so maybe it's up to them to check this.

Reply:
Hm, but then we have to carry around the types of the data every, I think that makes it more complicated and we might need GADTs as well, ie.

```
| ScriptTarget :: IsData a => ValidatorHash -> a -> Value -> EscrowTarget
```

Reply:
Also it might be nicer to just use the hash here to avoid storing the data again

Reply:
> Even different validators written in Plutus might decode Data objects in different ways, or treat them differently once they're decoded (there's no reason why something encoded with the List constructor can't be decoded into a random-aceesss array/vector, for example). Are we implicitly assuming that a Data "script" has a particular semantics which must be respected during validation?

Data has no semantics beyond its structure - the semantics in any given context is up to the validator script. How do you interoperate with this? Well, the author of the validator has to tell you what to do (i.e. what the semantics are in this case).

> Also it might be nicer to just use the hash here to avoid storing the data again

Didn't quite follow this?

Reply:
The current definition is `ScriptTarget ValidatorHash DataScript Value`, so the `DataScript` value appears here (in the data script of the transaction output locked by the contract) and on the output produced by the spending transaction. Since all we want to do with it is check whether it's identical, we could just have `ScriptTarget ValidatorHash DataScriptHash Value` here, and then compute the hash of the actual data script when the validator runs.

Reply:
Let's stick with equality on `Data`. If it turns out we need something else we can always add it later.

Reply:
Well, I guess I'm saying that it's weird: I'd have it be `ScriptTarget ValidatorHash PubKey Value`, and do things in the typed way. But it's okay.

Reply:
I see, got it. I'll make it a `Maybe DataScript` and then it's up to the user if they want to prescribe a specific value here or not.

Reply:
Actully that messes with `mkTxOut`, and the whole idea that you can construct a valid transaction that redeems the escrow outputs just from the `EscrowParams`.

Reply:
I don't get it - you just call `toData` to make the data script, no? Anyway, I'll just write you a patch and you can see if you like it.